nixos/tests/iscsi.nix

```
@@ -0,0 +1,94 @@
+import ./make-test.nix ({ lib, ... } :
+
+let
+  targetName = "iqn.2004-01.org.nixos:target";
+  initiatorName = "iqn.2004-01.org.nixos:initiator";
+
+in {
+  name = "iscsi";
+
+  meta = {
+    maintainers = [ lib.maintainers.markuskowa ];
+  };
+
+  nodes = {
+    server = { pkgs, ... } : {
+
+      boot.initrd.postDeviceCommands = ''
+        ${pkgs.e2fsprogs}/bin/mkfs.ext4 -L images /dev/vdb
+      '';
+
+      # One disk for FILEIO and one for BLOCKIO
+      virtualisation.emptyDiskImages = [ 1024 1024 ];
+
+      fileSystems = pkgs.lib.mkVMOverride
+        [ { mountPoint = "/images";
+            device = "/dev/disk/by-label/images";
+            fsType = "ext4";
+          }
+        ];
+
+      networking.firewall.allowedTCPPorts = [ 3260 ];
+      environment.systemPackages = [ pkgs.targetcli ];
+    };
+
+    client = { pkgs, lib, nodes, ... } : {
+      environment.systemPackages = with pkgs; [ e2fsprogs openiscsi parted ];
+
+      fileSystems = pkgs.lib.mkVMOverride [
+        { mountPoint = "/blockio";
+            device = "/dev/disk/by-label/blockio";
+            fsType = "ext4";
+            options = [ "_netdev" "noauto" ];
+        }
+        { mountPoint = "/fileio";
+            device = "/dev/disk/by-label/fileio";
+            fsType = "ext4";
+            options = [ "_netdev" "noauto" ];
+        }
+      ];
+    };
+  };
+
+  testScript = ''
+    $server->waitForUnit("multi-user.target");
+
+    # Directory must exist for targetcli to work
+    $server->succeed("mkdir -p /etc/target");
+
+    # Create stores
+    $server->succeed("targetcli backstores/fileio create file_or_dev=/images/image1 name=image1 size=500M");
+    $server->succeed("targetcli backstores/block create dev=/dev/vdc name=image2");
+
+    # Setup target
+    $server->succeed("targetcli iscsi/ create ${targetName}");
+    $server->succeed("targetcli iscsi/${targetName}/tpg1/acls/ create ${initiatorName}");
+
+    $server->succeed("targetcli iscsi/${targetName}/tpg1/luns create storage_object=/backstores/fileio/image1");
+    $server->succeed("targetcli iscsi/${targetName}/tpg1/luns create storage_object=/backstores/block/image2");
+
+    # Check client login
+    $client->waitForUnit("multi-user.target");
+    $client->succeed("iscsistart -i ${initiatorName} -t ${targetName} -g 1 -a server");
```

Maybe a race condition?

```suggestion
    $client->waitUntilSucceeds("iscsistart -i ${initiatorName} -t ${targetName} -g 1 -a server");
```

---

Thanks that seems to work.

---

Sorry, I got something wrong here. Still fails with ofBorg.

---

Is the command necessary after all or what does:

```
iscsistart: initiator reported error (15 - session exists)
```

mean?

---

`iscsistart` connects manually to the server (target). It is required to create new iSCSI provided block devices on the client ("/dev/sda" and "/dev/sdb" in this case). The message `session exists` means that the connect worked only half way at the first try (it somehow connected to target but still failed).